src/devices/cpu/z180/z180asci.cpp

```diff
@@ -0,0 +1,287 @@
+//  z180asci_channel - constructor
+//-------------------------------------------------
+
+z180asci_channel::z180asci_channel(
+		const machine_config &mconfig,
+		device_type type,
+		const char *tag,
+		device_t *owner,
+		uint32_t clock)
+	: device_t(mconfig, type, tag, owner, clock),
+	device_buffered_serial_interface(mconfig, *this),
+	m_tx_handler(*this)
+{
+	m_asci_cntla = 0;
+	m_asci_cntlb = 7; // SS2,SS1 and SS0 are 1 after reset
+	m_asci_stat = 2; // set TDRE=1 on reset
+	m_asci_tdr = 0;
+	m_asci_rdr = 0;
+	m_id = 0;
+	m_irq = 0;
+}
+
+z180asci_channel::z180asci_channel(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: z180asci_channel(mconfig, Z180ASCI_CHANNEL, tag, owner, clock)
+{
+}
+
+//-------------------------------------------------
+//  device_resolve_objects - resolve objects that
+//  may be needed for other devices to set
+//  initial conditions at start time
+//-------------------------------------------------
+
+void z180asci_channel::device_resolve_objects()
+{
+	// resolve callbacks
+	m_tx_handler.resolve_safe();
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void z180asci_channel::device_start()
+{
+	save_item(NAME(m_asci_cntla));
+	save_item(NAME(m_asci_cntlb));
+	save_item(NAME(m_asci_stat));
+	save_item(NAME(m_asci_tdr));
+	save_item(NAME(m_asci_rdr));
+
+	transmit_register_reset();
+	receive_register_reset();
+
+	cpu_device *cpu = reinterpret_cast<cpu_device *>(owner());
+	if (m_id==0) {
+		cpu->state_add(Z180_CNTLA0,     "CNTLA0",    m_asci_cntla);
+		cpu->state_add(Z180_CNTLB0,     "CNTLB0",    m_asci_cntlb);
+		cpu->state_add(Z180_STAT0,      "STAT0",     m_asci_stat);
+		cpu->state_add(Z180_TDR0,       "TDR0",      m_asci_tdr);
+		cpu->state_add(Z180_RDR0,       "RDR0",      m_asci_rdr);
+	} else {
+		cpu->state_add(Z180_CNTLA1,     "CNTLA1",    m_asci_cntla);
+		cpu->state_add(Z180_CNTLB1,     "CNTLB1",    m_asci_cntlb);
+		cpu->state_add(Z180_STAT1,      "STAT1",     m_asci_stat);
+		cpu->state_add(Z180_TDR1,       "TDR1",      m_asci_tdr);
+		cpu->state_add(Z180_RDR1,       "RDR1",      m_asci_rdr);
+	}
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void z180asci_channel::device_reset()
+{
+	m_asci_cntla = 0;
+	m_asci_cntlb = 7; // SS2,SS1 and SS0 are 1 after reset
+	m_asci_tdr = 0;
+	m_asci_rdr = 0;
+	if (m_id==0) {
+		m_asci_cntla = (m_asci_cntla & Z180_CNTLA0_MPBR_EFR) | Z180_CNTLA0_RTS0;
+		m_asci_cntlb = (m_asci_cntlb & (Z180_CNTLB0_MPBT | Z180_CNTLB0_CTS_PS)) | 0x07;
+		m_asci_stat = m_asci_stat & (Z180_STAT0_DCD0 | Z180_STAT0_TDRE);
+	} else {
+		m_asci_cntla = (m_asci_cntla & Z180_CNTLA1_MPBR_EFR) | Z180_CNTLA1_CKA1D;
+		m_asci_cntlb = (m_asci_cntlb & Z180_CNTLB1_MPBT) | 0x07;
+		m_asci_stat = Z180_STAT1_TDRE;
+	}
```

Comment:
The whole switching behaviour on ID doesn’t seem right.  Can you configure the devices as necessary when the parent device instantiates them?  Are the two of them sufficiently different that they should be different device types?

Reply:
They are diffenet in a way that port 0 have more capabilities then port 1. Also z8s180 variant have additional features